### ISSUE TITLE
Juntao - Users may not be able to fully redeem USD1 into USDT even when reserve ratio is above 100%

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-04-unitasprotocol-judging#95]
Juntao

high

# Users may not be able to fully redeem USD1 into USDT even when reserve ratio is above 100%

## Summary

Users may not be able to fully redeem USDT even when reserve ratio is above 100%, because of portfolio being taken into the account for calculation.

## Vulnerability Detail

Reserve ratio shows how many liabilities is covered by reserves, a reserve ratio above 100% guarantees protocol has enough USDT to redeem, the way of calculating reserve ratio is `Reserve Ratio = allReserves / liabilities` and is implemented in [Unitas#_getReserveStatus(...)](https://github.com/sherlock-audit/2023-04-unitasprotocol/blob/main/Unitas-Protocol/src/Unitas.sol#L476-L495) function:
```solidity
            reserveRatio = ScalingUtils.scaleByBases(
                allReserves * valueBase / liabilities,
                valueBase,
                tokenManager.RESERVE_RATIO_BASE()
            );
```

`allReserves` is the sum of the [balance](https://github.com/sherlock-audit/2023-04-unitasprotocol/blob/main/Unitas-Protocol/src/PoolBalances.sol#L22) of Unitas and InsurancePool,  calculated in [Unitas#_getTotalReservesAndCollaterals()](https://github.com/sherlock-audit/2023-04-unitasprotocol/blob/main/Unitas-Protocol/src/Unitas.sol#L500-L535) function:
```solidity
        for (uint256 i; i < tokenCount; i++) {
            address token = tokenManager.tokenByIndex(tokenTypeValue, i);
            uint256 tokenReserve = _getBalance(token);
            uint256 tokenCollateral = IInsurancePool(insurancePool).getCollateral(token);


            if (tokenReserve > 0 || tokenCollateral > 0) {
                uint256 price = oracle.getLatestPrice(token);


                reserves += _convert(
                    token,
                    baseToken,
                    tokenReserve,
                    MathUpgradeable.Rounding.Down,
                    price,
                    priceBase,
                    token
                );


                collaterals += _convert(
                    token,
                    baseToken,
                    tokenCollateral,
                    MathUpgradeable.Rounding.Down,
                    price,
                    priceBase,
                    token
                );
            }
        }
```

`liabilities` is the total value of USD1 and USDEMC tokens, calculated in [Unitas#_getTotalLiabilities()](https://github.com/sherlock-audit/2023-04-unitasprotocol/blob/main/Unitas-Protocol/src/Unitas.sol#L540-L567) function:
```solidity
        for (uint256 i; i < tokenCount; i++) {
            address token = tokenManager.tokenByIndex(tokenTypeValue, i);
            uint256 tokenSupply = IERC20Token(token).totalSupply();


            if (token == baseToken) {
                // Adds up directly when the token is USD1
                liabilities += tokenSupply;
            } else if (tokenSupply > 0) {
                uint256 price = oracle.getLatestPrice(token);


                liabilities += _convert(
                    token,
                    baseToken,
                    tokenSupply,
                    MathUpgradeable.Rounding.Down,
                    price,
                    priceBase,
                    token
                );
            }
        }
```

Some amount of USDT in both Unitas and InsurancePool is [portfolio](https://github.com/sherlock-audit/2023-04-unitasprotocol/blob/main/Unitas-Protocol/src/PoolBalances.sol#L26), which represents the current amount of assets used for strategic investments, it is worth noting that after [sending portfolio](https://github.com/sherlock-audit/2023-04-unitasprotocol/blob/main/Unitas-Protocol/src/PoolBalances.sol#L83-L99), `balance` remains the same, which means `portfolio` is taken into account in the calculation of reserve ratio.

This is problematic because `portfolio` is not available when user redeems, and user may not be able to fully redeem for USDT even when protocols says there is sufficient reserve ratio. 

Let's assume :

> Unitas's balance is 10000 USD and its portfolio is 2000 USD, avaliable balance is 8000 USD
> InsurancePool's balance is 3000 USD and its portfolio is 600 USD, available balance is 2400 USD
> AllReserves value is 13000 USD
> Liabilities (USDEMC) value is 10000 USD
> Reserve Ratio is (10000 + 3000) /  10000 = 130%.

Later on, USDEMC appreciates upto 10% and we can get:
> AllReserves value is still 13000 USD
> Liabilities (USDEMC) value is 11000 USD
> Reserve Ratio is (10000 + 3000) /  11000 = 118%.

The available balance in Unitas is 8000 USD so there is 3000 USD in short, it needs to be obtain from InsurancePool, however,  the available balance in InsurancePool is 2400 USD, transaction will be reverted and users cannot redeem.

There would also be an extreme situation when reserve ratio is above 100% but there is no available balance in protocol because all the `balance` is `portfolio` (this is possible when InsurancePool is drained out), users cannot redeem any USDT in this case.

## Impact

Users may not be able to fully redeem USD1 into USDT even when reserve ratio is above 100%, this defeats the purpose of reserve ratio and breaks the promise of the protocol, users may be mislead and lose funds.

## Code Snippet

https://github.com/sherlock-audit/2023-04-unitasprotocol/blob/main/Unitas-Protocol/src/Unitas.sol#L500-L535

## Tool used

Manual Review

## Recommendation

Portfolio should not be taken into account for the calculation of reserve ratio.
```diff
    function _getTotalReservesAndCollaterals() internal view returns (uint256 reserves, uint256 collaterals) {
        ...
-           uint256 tokenReserve = _getBalance(token);
+           uint256 tokenReserve = _getBalance(token) - _getPortfolio(token);
-           uint256 tokenCollateral = IInsurancePool(insurancePool).getCollateral(token);
+           uint256 tokenCollateral = IInsurancePool(insurancePool).getCollateral(token) - IInsurancePool(insurancePool).getPortfolio(token);
        ...
    }
```

**Comment by 0xruhum** [comment]
Escalate

This is not just a design choice. There are no specifics on what a "small portion" or "enough capital in reserve and insurance pool" means. Allowing the protocol team to remove collateral from the system to earn yield is a security risk. In the current version, USDT is the collateral for USD1. Without USDT, USD1 is worth nothing. The reserve ratio is used to keep the protocol healthy at all times. By including the funds inside the portfolio in the calculation you allow the protocol to become under-collateralized under certain circumstances.

Considering that #16 describes the risk of the portfolio incurring a loss which is deemed a valid issue, it's reckless to account for the funds inside the portfolio when calculating the reserve ratio. From a protocol perspective, these funds are not part of the system anymore. They were moved into a different protocol to earn yield. That is fine as long as you keep the reserve ratio of Unitas above 130%. For that to be the case, the funds inside the portfolio have to be **excluded** from the reserve ratio calculation.

To sum up, any funds inside the portfolio should not be used to calculate the reserve ratio. They can be gone at any moment due to the risks described in #16.

**Comment by thangtranth** [comment]
Escalate for 10 USDC

Reading the original report, I see no mentioning of the potential path or risk that leads to the issue. However in the escalation, the arguments are built upon the insights and finding of another report after submission. If the original report stands alone, I am not sure that it is convincing enough.

This escalation also aims to clarify the rules for other contestants who may do the same in the future.

**Comment by jacksanford1** [comment]
@0xruhum I see what you're getting at. But the only way this can be a valid issue is if the protocol has intended for users to always be able to redeem USD1 into USDT when the reserve ratio is 100% or greater. I don't think Unitas has made that claim anywhere? If you see a claim like this (in comments or in code), please show me.

**Comment by jacksanford1** [comment]
Update @0xruhum: Based on the following sentence in the README, the broader point of this issue should be true:

> The Unitas protocol guarantees unrestricted and unconditional conversion of its unitized stablecoins “back” to USD-pegged stablecoins.

The issue correct gets that "users may not be able to fully redeem USD1 into USDT even when the reserve ratio is above 100%" so this issue should be a valid Medium (maybe even High). But the portfolio aspect is not super unique, so it should be duplicated with any other issues that get the general idea of this vulnerability.

**Comment by jacksanford1** [comment]
Escalation accepted

Duplicate of #13

**Comment by 0xruhum** [comment]
> But the portfolio aspect is not super unique, so it should be duplicated with any other issues that get the general idea of this vulnerability.

Yep. Just wanted to mention again that the issues I've linked in the original escalation comment aren't valid duplicates tho. That should be taken into account.

**Comment by jacksanford1** [comment]
Ok @0xruhum, as far as I'm aware none of them are considered to be duplicates right now.

**Comment by Jiamincoin** [comment]
> Ok @0xruhum, as far as I'm aware none of them are considered to be duplicates right now.

Hey I think #114 should be taken into account too. 
I didn't create escalation because it was considered duplicate of the main report in the first place.

**Comment by jacksanford1** [comment]
Ok @Jiamincoin, I agree that #114 is a duplicate of this issue.

**Comment by 0x00ffDa** [comment]
I do not see any discussion of duplicate [#118](https://github.com/sherlock-audit/2023-04-unitasprotocol-judging/issues/118).  Please consider it before ending the escalation evaluations.

**Comment by jacksanford1** [comment]
Result:
Medium
Has Duplicates

**Comment by hrishibhat** [comment]
Result:
Medium
Has duplicates

**Comment by sherlock-admin** [comment]
Escalations have been resolved successfully!

Escalation status:
- [0xruhum](https://github.com/sherlock-audit/2023-04-unitasprotocol-judging/issues/95/#issuecomment-1603137490): accepted
- [thangtranth](https://github.com/sherlock-audit/2023-04-unitasprotocol-judging/issues/95/#issuecomment-1605657584): rejected

**Comment by jacksanford1** [comment]
Acknowledged by protocol team (won't fix).
